### PR TITLE
[wip] [runtime] Fix scanning of SIMD registers on Mac.

### PR DESCRIPTION
**Pull request by evincarofautumn**
Testing the Mac parts of #3364 in isolation.

### CODE REVIEW

**Comment by evincarofautumn** [comment]
build

**Comment by evincarofautumn** [comment]
build

**Comment by akoeplinger** [comment]
This breaks the maccore build:

```
[7:35:59] /Users/builder/data/lanes/1411/0760eeaf/source/xamarin-macios/external/mono/mono/metadata/sgen-stw.c:71:2: error: implicit declaration of function 'mono_mach_arch_thread_state_to_mono_context' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
[7:35:59]         MONO_CONTEXT_GET_CURRENT (info->client_info.ctx);
[7:35:59]         ^
[7:35:59] /Users/builder/data/lanes/1411/0760eeaf/source/xamarin-macios/external/mono/mono/utils/mono-context.h:408:2: note: expanded from macro 'MONO_CONTEXT_GET_CURRENT'
[7:35:59]         mono_mach_arch_thread_state_to_mono_context ((thread_state_t)&thread_state, &ctx); \
[7:35:59]         ^
[7:35:59] /Users/builder/data/lanes/1411/0760eeaf/source/xamarin-macios/external/mono/mono/metadata/sgen-stw.c:71:2: note: did you mean 'mono_mach_arch_thread_states_to_mono_context'?
[7:35:59] /Users/builder/data/lanes/1411/0760eeaf/source/xamarin-macios/external/mono/mono/utils/mono-context.h:408:2: note: expanded from macro 'MONO_CONTEXT_GET_CURRENT'
[7:35:59]         mono_mach_arch_thread_state_to_mono_context ((thread_state_t)&thread_state, &ctx); \
[7:35:59]         ^
[7:35:59] /Users/builder/data/lanes/1411/0760eeaf/source/xamarin-macios/external/mono/mono/utils/mach-support.h:34:6: note: 'mono_mach_arch_thread_states_to_mono_context' declared here
[7:35:59] void mono_mach_arch_thread_states_to_mono_context (thread_state_t state, thread_state_t fpstate, MonoContext *context);
[7:35:59]      ^
[7:35:59] 1 error generated.
[7:35:59] make[4]: *** [libmonoruntimesgen_la-sgen-stw.lo] Error 1
[7:35:59] make[3]: *** [all-recursive] Error 1
[7:35:59] make[2]: *** [build-targettv] Error 2
```

**Comment by evincarofautumn** [comment]
Probably because of the removal of `sgen-os-mach.c`.

**Comment by vargaz** [comment]
This also causes runtime test failures:
https://wrench.internalx.com/Wrench/index.aspx?limit=50
Specially, the bug-45841-fpstack-exceptions.exe test fails 100% of the time.

File: mono/utils/mach-support-amd64.c [hunk 1]
```diff
@@ -104,23 +123,39 @@ mono_mach_arch_get_thread_state_size ()
...
 kern_return_t
-mono_mach_arch_set_thread_state (thread_port_t thread, thread_state_t state, mach_msg_type_number_t count)
+mono_mach_arch_set_thread_states (thread_port_t thread, thread_state_t state, mach_msg_type_number_t count, thread_state_t fpstate, mach_msg_type_number_t fpcount)
 {
-	return thread_set_state (thread, x86_THREAD_STATE64, state, count);
+	kern_return_t ret;
+	ret = thread_set_state (thread, x86_THREAD_STATE64, state, count);
+	if (ret != KERN_SUCCESS)
+		return ret;
+	ret = thread_set_state (thread, x86_FLOAT_STATE64, fpstate, fpcount);
```
Comment (kumpera) [inline mono/utils/mach-support-amd64.c]: Why are we restoring the FP state?

**Pull request merged**